arch/x86/gen_gdt.py

```diff
@@ -214,6 +221,15 @@ def main():
             fp.write(create_code_data_entry(0, 0xFFFFF, 3,
                                             FLAGS_GRAN, ACCESS_RW))
 
+        if use_tls:
+            # Selector 0x18, 0x28 or 0x38 (depending on entries above):
+            # data descriptor, dpl = 3
+            #
+            # for use with thread local storage while this will be
+            # modified at runtime.
+            fp.write(create_code_data_entry(0, 0xFFFFF, 3,
+                                            FLAGS_GRAN, ACCESS_RW))
+
```

Comment:
The length of this segment could be set to the length of the TLS area for a bit of additional protection.

Reply:
The limit here should not matter, as data access permission is done through `%ds`. Only the base address in GS is being used.

Reply:
This confuses me. I thought that the %gs limit took effect whenever %gs was used to access memory. Setting the %gs limit should cause a fault if you access outside of the TLS area
```
__thread int x[10];
int foo(void) { return x[12]; }
```
which compiles into this:
```
   0x11e0 <foo>:	mov    %gs:0x8,%eax
   0x11e6 <foo+6>:	ret    
```
If we placed the %gs segment in 'expand down' mode (type 7), and set the limit to be something like -(TLS size), then I think we'd get a trap when this program executed. Of course, taking the address of a TLS variable and accessing that would still use %ds. Given that the TLS size is known at compile time, this *seems* like it would be "free" to add (no additional code or runtime imposed).

Reply:
even without a limit, if a user thread accesses outside of its TLS region past the bounds of its stack buffer it will get a page fault.
a supervisor thread messing with strange %gs offsets, I suspect we generally don't care much.
I don't think there's any security concerns here, at least?

Reply:
`foo()` compiles to this under Zephyr:
```
push %ebp
mov  %esp, %ebp
mov  %gs:0x0, %eax
mov  0x8(%eax), %eax
pop  %ebp
ret
```
So this would be governed by `%ds` instead.